### PR TITLE
Fix "405 METHOD NOT ALLOWED" by (1) Removing query parameters (2) Changing /media to /meta

### LINKED ISSUE #106: Videos sometimes fail to "Download to IIAB" w/ message: 405 METHOD NOT ALLOWED
Example video that failed for @avni: (in a Multipass 1.13.0 VM on Mac)

"India-Pakistan partition explained"
https://www.youtube.com/watch?v=OnTYLyNUPMc

image

iiab-diagnostics:

- https://sprunge.us/Ujlsn8?en
- https://sprunge.us/Ujlsn8?en#n-805

@deldesir: are these 2 issues below related or no?

- #85 
- #101

Tangentially related, also from @avni:

- #77

### PR DESCRIPTION
**Pull request by deldesir**
🚀 **Pull Request Overview:**
Fixes 405 METHOD NOT ALLOWED by cleaning up the endpoint so no query parameters are included and also ensure its correctness (/meta) when the download button is clicked outside of the tasks view.

📋 **Checklist:**
- [x] Tested the change on 10.8.0.22 (Ubuntu 24.04)

Related issue: #106

### CODE REVIEW

**Comment by deldesir** [comment]
@EMG70, to test this PR effectively, download any video from within any page of Calibre-Web.

File: cps/tasks/metadata_extract.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def __init__(self, task_message, media_url, original_url, current_user_name):
         self.message = task_message
         self.media_url = media_url
         self.media_url_link = f'<a href="{media_url}" target="_blank">{media_url}</a>'
-        self.original_url = original_url
+        self.original_url = original_url.split("?")[0].replace("media", "meta")
```
Comment (holta) [inline cps/tasks/metadata_extract.py:22]: Can regex be made safer?

(Please test things like... ?)

```suggestion
        self.original_url = original_url.split("?")[0].replace("/media$", "/meta")
```

**Comment by EMG70** [comment]
> @EMG70, to test this PR effectively, download any video from within any page of Calibre-Web.
IIAB-DIAGNOSTICS  http://sprunge.us/XZ87Su?en
Download of playlist https://www.youtube.com/playlist?list=PLpQQipWcxwt_wKeFEmZL15qOZEkiVUQAq
Screenshot from 2024-02-21 18-07-18

Below is a download of individual video from calibre (already downloaded as part of above playlist) https://www.youtube.com/watch?v=5mMAGcZaItg
Screenshot from 2024-02-21 18-07-44

No issues found,all working as described after pull request.

**Comment by deldesir** [comment]
Thanks for testing

File: cps/tasks/metadata_extract.py (hunk 1 above)
Comment (deldesir) [reply cps/tasks/metadata_extract.py:22]: Definitely, but why "$"?

Comment (holta) [reply cps/tasks/metadata_extract.py]: Nobody wants corner cases, perpetually exploding in our face! 🤯

So please emphasize code safety — verifying `/media` on the right side of the input string — here's 1 way of doing that:

https://stackoverflow.com/questions/3675318/how-to-replace-some-characters-from-the-end-of-a-string/3675401#3675401

Comment (holta) [reply cps/tasks/metadata_extract.py]: > verifying `/media` on the right side of the input string

(Or whatever precondition... that we should be testing for!)

**Comment by holta** [comment]
Update from Thursday.  I've asked @deldesir to validate string manipulations properly going forward:

- "Quick & Dirty" (hand-waving, wishful thinking) are just no longer good enough! ❌

- [re.sub](https://stackoverflow.com/questions/3675318/how-to-replace-some-characters-from-the-end-of-a-string/3675401#3675401) (proper regex, or similar) always need to be used.  So that bugs don't explode in face (and everyone's face) down the road.  RECAP: Validate all assumptions a.k.a. expected preconditions — by enforcing them with code (if a substring is expected to be on the right-side of a string, enforce that!) ✅

**Comment by EMG70** [comment]
IIAB-DIAGNOSTICS http://sprunge.us/lBgw9H?en
 A new VM was installed after PR#131 was emerged .No issues were observed with redownloading.The 405 error has not been encountered this time.
A playlist https://www.youtube.com/playlist?list=PLpQQipWcxwt_wKeFEmZL15qOZEkiVUQAq was downloaded twice and individual videos https://www.youtube.com/watch?v=5mMAGcZaItg from within calibre also downloaded OK.
Screenshot from 2024-02-24 18-08-11

**Comment by holta** [comment]
Profound thanks @EMG70. ✅

Hopefully @avni might also time — to try to reproduce the same conditions that originally led to this problem — as she reported on Jan 20:

- #106

**Comment by EMG70** [comment]
> Profound thanks @EMG70. ✅
> 
> Hopefully @avni might also time — to try to reproduce the same conditions that originally led to this problem — as she reported on Jan 20:
> 
>     * [Videos sometimes fail to "Download to IIAB" w/ message: 405 METHOD NOT ALLOWED #106](https://github.com/iiab/calibre-web/issues/106)

Another test with  @avni's previously failed video  https://www.youtube.com/watch?v=OnTYLyNUPMc did not return 405 failed error or any errors.
Screenshot from 2024-02-24 18-27-53

**Comment by holta** [comment]
> Another test with @avni's previously failed video https://www.youtube.com/watch?v=OnTYLyNUPMc did not return 405 failed error or any errors.

Awesome news!

And hopefully she too can reconfirm, on macOS under her conditions, to be 100% sure.

**Comment by avni** [comment]
Confirmed! I tried from both http://192.168.64.17/books/ and http://192.168.64.16/books/tasks and it now works. Thank you!

**Pull request merged**